luigi_monitor/luigi_monitor.py

```diff
@@ -72,37 +88,83 @@ def set_handlers(EVENTS):
         function = event_map[event]['function']
         luigi.Task.event_handler(handler)(function)
 
+
+def slack_attachment(text, color):
+    """See https://api.slack.com/docs/message-attachments (for our luigi-monitor status messages)"""
+    return {
+        "text": text,
+        "color": color,
+        "fields": [{
+            "title": None,
+            "value": None,
+            "short": False
+        }]
+    }
+
+
 def format_message(max_print, job):
-    text = ["Status report for {}".format(job)]
+    job_status_report_title = "Status report for {}".format(job)
+    slack_message_payload_attachments = {"attachments": []}
+
     if 'Success' in EVENTS:
-        text.append("*Successes:*")
-        if len(EVENTS['Success']) > max_print:
-            text.append("More than %d Successes. Please check logs." % max_print)
-        else:
-            for success in EVENTS['Success']:
-                text.append("Task: {}".format(success['task']))
+        success_attachment = get_success_slack_attachment(max_print)
+        slack_message_payload_attachments['attachments'].append(success_attachment)
+
     if 'Failure' in EVENTS:
-        text.append("*Failures:*")
-        if len(EVENTS['Failure']) > max_print:
-            text.append("More than %d failures. Please check logs." % max_print)
-        else:
-            for failure in EVENTS['Failure']:
-                text.append("Task: {}; Exception: {}".format(failure['task'], failure['exception']))
+        failure_attachment = get_failure_slack_attachment(max_print)
+        slack_message_payload_attachments['attachments'].append(failure_attachment)
+
     if 'Missing' in EVENTS:
-        text.append("*Tasks with missing dependencies:*")
-        if len(EVENTS['Missing']) > max_print:
-            text.append("More than %d tasks with missing dependencies. Please check logs." % max_print)
-        else:
-            for missing in EVENTS['Missing']:
-                text.append(missing)
-    if len(text) == 1:
-        text.append("Job ran successfully!")
-    text = "\n".join(text)
-    return text
+        missing_attachment = get_missing_dependencies_slack_attachement(max_print)
+        slack_message_payload_attachments['attachments'].append(missing_attachment)
+
+    return job_status_report_title, slack_message_payload_attachments
+
+
+def get_missing_dependencies_slack_attachement(max_print):
```

Comment:
These three methods are all very similar!

- Think DRY (don't repeat yourself) and Open-Closed (what if we wanted to add a slack notification for a new event, e.g. not just missing, failure, success)

Ideally (and it won't take long) I would create a `SlackNotifications` class

- you can have a list of events it can handle ('Success', 'Failure', 'Missing') and loop over them
return the slack_message_payload

- you can have a `get_attachment` which is generic (which takes an event type)
- the event specific formatting can be handled by private methods.

Let me know if you want to sketch out this class in more detail on paper - and review it with me. Then you can implement.